Uninstalling a plugin #1531 (#1538)
jenkinsci/docker: README.md

Otherwise a banner will appear prompting the user to install additional plugins,
which may be inappropriate.

# Uninstalling Deprecated Plugins in Jenkins (with Docker)

This guide provides detailed instructions on how to uninstall deprecated plugins in Jenkins when using Docker. Removing deprecated plugins is crucial to maintain the security and stability of your Jenkins instance.

## Prerequisites

- Docker and Docker Compose are installed on your system.
- Basic knowledge of Docker and Jenkins.

## Steps

1. **Locate `plugins.txt`**: Find the `plugins.txt` file in your Jenkins Docker setup. This file defines the plugins installed in your Jenkins instance. It can be found in the Dockerfile or a related configuration directory.

Example content of `plugins.txt`:

```plaintext
# See https://github.com/jenkinsci/docker#usage-1
ant:1.11
bootstrap4-api:4.6.0-3
popper-js:2.9.2-1
```

> 2. **Update plugins.txt**: Open the plugins.txt file and identify the plugins that are deprecated. These plugins are indicated by the deprecation warning or message in the Jenkins UI. Remove the lines corresponding to the deprecated plugins from the plugins.txt file. Save the changes.

Reviewer: ```suggestion
2. **Update plugins.txt**: Open the `plugins.txt` file and identify the plugins that are deprecated.
These plugins are indicated by the deprecation warning or message in the Jenkins UI.
Remove the lines corresponding to the deprecated plugins from the `plugins.txt` file.
Save the changes.
```

> Example updated plugins.txt (with deprecated plugins removed):

Reviewer: ```suggestion
Example updated `plugins.txt` (with deprecated plugins removed):
```

```plaintext
# See https://github.com/jenkinsci/docker#usage-1
ant:1.11
```

> 3. **Rebuild the Jenkins Image**: In your terminal, navigate to the directory containing the Docker configuration files for your Jenkins instance. Rebuild the Jenkins image to apply the changes made in the plugins.txt file using the following command:

Reviewer: ```suggestion
3. **Rebuild the Jenkins Image**: In your terminal, navigate to the directory containing the Docker configuration files for your Jenkins instance.
Rebuild the Jenkins image to apply the changes made in the plugins.txt file using the following command:
```

> ```
> docker-compose build jenkins
> ```

Reviewer: ```suggestion
```bash
docker compose build jenkins
```
```

> 4. Recreate the Jenkins Container: Once the image is successfully rebuilt, recreate the Jenkins container using the updated image. Use the following command:

Reviewer: ```suggestion
4. Recreate the Jenkins Container: Once the image is successfully rebuilt, recreate the Jenkins container using the updated image.
Use the following command:
```

> ```
> docker-compose up -d --build --force-recreate jenkins
> ```

Reviewer: ```suggestion
```bash
docker compose up -d --build --force-recreate jenkins
```
```

I'm not so sure the `--build` is useful at this stage.

5. Verify Plugin Removal: To verify if the deprecated plugins have been removed, enter the running Jenkins container by executing the following command:

> ```
> docker-compose exec jenkins bash
> ```

Reviewer: ```suggestion
```bash
docker compose exec jenkins bash
```
```

6. Once inside the container, navigate to the Jenkins plugin directory:

> ```
> cd /usr/share/jenkins/
> ```

Reviewer: ```suggestion
```bash
cd /usr/share/jenkins/
```
```

> 7. Check the plugins.txt file to ensure the references to the deprecated plugins are gone:

Reviewer: ```suggestion
7. Check the `plugins.txt` file to ensure the references to the deprecated plugins are gone:
```

> ```
> cat plugins.txt | grep <plugin-name>
> ```

Reviewer: ```suggestion
```bash
cat plugins.txt | grep <plugin-name>
```
```

> Replace <plugin-name> with the name of the deprecated plugin you want to uninstall. If no results are returned, it means the plugin references have been removed successfully.

Reviewer: ```suggestion
Replace `<plugin-name>` with the name of the deprecated plugin you want to uninstall.
If no results are returned, it means the plugin references have been removed successfully.
```

> 8. Uninstall Plugins in the Jenkins UI: Access the Jenkins web interface by visiting http://localhost:8080 in your web browser. Log in with your admin credentials.

Reviewer: ```suggestion
8. Uninstall Plugins in the Jenkins UI: Access the Jenkins web interface by visiting [http://localhost:8080](`http://localhost:8080`) in your web browser.
Log in with your admin credentials.
```

> 9. Navigate to Manage Jenkins and select Manage Plugins.

Reviewer: ```suggestion
9. Navigate to _Manage Jenkins_ and select _Manage Plugins_.
```

> 10. In the Installed tab, search for the name of the deprecated plugin you want to uninstall.

Reviewer: ```suggestion
10. In the _Installed_ tab, search for the name of the deprecated plugin you want to uninstall.
```

> 11. If the plugin appears in the list, click the Uninstall button (red cross) next to it. If the button is disabled then it means that the plugin has a dependency on another plugin, just hover on the red cross to know the parent plugin and search that plugin and remove it first.

Reviewer: ```suggestion
11. If the plugin appears in the list, click the _Uninstall_ button (red cross) next to it.
If the button is disabled then it means that the plugin has a dependency on another plugin.
Just hover over the red cross to know the parent plugin.
Search that plugin and remove it first by following the same procedure.
```

12. Confirm the removal when prompted.

> 13. Restart Jenkins: To complete the removal process, Jenkins needs to be restarted. Restart Jenkins by hitting the
>     `/safeRestart` endpoint or using the Jenkins UI.

Reviewer: ```suggestion
13. Restart Jenkins: To complete the removal process, Jenkins needs to be restarted.
Restart Jenkins by hitting the `/safeRestart` endpoint or using the Jenkins UI.
```

> 14. Optional: Remove Plugin Remnants from Docker Volume: If you want to remove any remnants of the deprecated plugins from the Docker volume, follow these steps:

Reviewer: ```suggestion
14. Optional: Remove Plugin remnants from Docker Volume: If you want to remove any remnants of the deprecated plugins from the Docker volume, follow these steps:
```

- Enter the running Jenkins container using the command mentioned in step 5.
- Navigate to the plugin directory:

`cd ~/plugins`

- List the contents of the directory to identify the remnants of the deprecated plugins:

`ls -artl *`

- Remove the remnants of the deprecated plugins using the rm command. For example:

`rm -fr <plugin-name>*`

> Replace <plugin-name> with the name of the deprecated plugin you want to remove.

Reviewer: ```suggestion
Replace `<plugin-name>` with the name of the deprecated plugin you want to remove.
```

Otherwise a banner will appear prompting the user to install additional plugins,
which may be inappropriate.

# Uninstalling Deprecated Plugins in Jenkins (with Docker)

This guide provides detailed instructions on how to uninstall deprecated plugins in Jenkins when using Docker.
Removing deprecated plugins is crucial to maintain the security and stability of your Jenkins instance.

## Prerequisites

- Docker and Docker Compose are installed on your system.

> - Basic knowledge of Docker and Jenkins.

Reviewer 2: This section does not make sense for me: this pull request is not expected to add a step by step guide that would only cover some use cases.

Do you mind replacing it by a summary of the main important steps instead?

additionnally, I do not expect this to be a guide using docker compose: not only `docker-compose` is deprecated in favor of the `docker compose` plugin (ref. https://docs.docker.com/compose/, also refered as "Docker Compose v2) but also the instructions in this README should be minimalistic to ensure they benefit non Docker users.

## Steps

1. **Locate `plugins.txt`**: Find the `plugins.txt` file in your Jenkins Docker setup.
   This file defines the plugins installed in your Jenkins instance.
   It can be found in the Dockerfile or a related configuration directory.

Example content of `plugins.txt`:

```plaintext
# See https://github.com/jenkinsci/docker#usage-1
ant:1.11
bootstrap4-api:4.6.0-3
popper-js:2.9.2-1
```

2. **Update plugins.txt**: Open the `plugins.txt` file and identify the plugins that are deprecated.
   These plugins are indicated by the deprecation warning or message in the Jenkins UI.
   Remove the lines corresponding to the deprecated plugins from the `plugins.txt` file.
   Save the changes.

Example updated `plugins.txt` (with deprecated plugins removed):

```plaintext
# See https://github.com/jenkinsci/docker#usage-1
ant:1.11
```

> 3. **Rebuild the Jenkins Image**: In your terminal, navigate to the directory containing the Docker configuration files for your Jenkins instance.

Reviewer 2: ```suggestion
3. **Rebuild the Jenkins Container Image**: In your terminal, navigate to the directory containing the Jenkins Container Image definition.
```

>    Rebuild the Jenkins image to apply the changes made in the plugins.txt file using the following command:
> 
> ```bash
> docker-compose build jenkins
> ```

Reviewer 2: ```suggestion
   Rebuild the Jenkins image to apply the changes made in the plugins.txt file using the appropriate tool for your setup (`docker build <...>`, `docker compose build <...>`, etc.).
```

- Remove Docker Compose

4. Recreate the Jenkins Container: Once the image is successfully rebuilt, recreate the Jenkins container using the updated image.

>    Use the following command:
> 
> ```bash
> docker-compose up -d --build --force-recreate jenkins
> ```

Reviewer 2: ```suggestion
```

Remove Docker Compose

> 5. Verify Plugin Removal: To verify if the deprecated plugins have been removed, enter the running Jenkins container by executing the following command:
> 
> ```bash
> docker-compose exec jenkins bash
> ```
> 
> 6. Once inside the container, navigate to the Jenkins plugin directory:
> 
> ```bash
> cd /usr/share/jenkins/
> ```
> 
> 7. Check the `plugins.txt` file to ensure the references to the deprecated plugins are gone:
> 
> ```bash
> cat plugins.txt | grep <plugin-name>
> ```
> 
> Replace `<plugin-name>` with the name of the deprecated plugin you want to uninstall.
> If no results are returned, it means the plugin references have been removed successfully.

Reviewer 2: ````suggestion
5. Verify effective plugins removal from the image: To verify if the deprecated plugins have been removed from the image, check their absence from the `/usr/share/jenkins/plugins.txt` in the running Jenkins container.

For instance, if you use `docker`:

```bash
$ docker exec <Jenkins Container ID> grep -c "<plugin-name>" /usr/share/jenkins/plugins.txt
0 # Should be zero (e.g. the number of occurences of the string "<plugin-name>" in the file)
```

Replace `<plugin-name>` with the name of the deprecated plugin you want to uninstall.
````

Proposal:

- Squash steps into a single one
- No need to use interactive steps
- A bit of rephrasing
- Remove Docker Compose

11. If the plugin appears in the list, click the _Uninstall_ button (red cross) next to it.
    If the button is disabled then it means that the plugin has a dependency on another plugin.
    Just hover on the red cross to know the parent plugin.
    Search that plugin and remove it first by following the same procedure.

12. Confirm the removal when prompted.

13. Restart Jenkins: To complete the removal process, Jenkins needs to be restarted.
    Restart Jenkins by hitting the `/safeRestart` endpoint or using the Jenkins UI.

14. Optional: Remove Plugin remnants from Docker Volume: If you want to remove any remnants of the deprecated plugins from the Docker volume, follow these steps:

- Enter the running Jenkins container using the command mentioned in step 5.
- Navigate to the plugin directory:

`cd ~/plugins`

- List the contents of the directory to identify the remnants of the deprecated plugins:

`ls -artl *`

- Remove the remnants of the deprecated plugins using the rm command. For example:

`rm -fr <plugin-name>*`

Replace `<plugin-name>` with the name of the deprecated plugin you want to remove.

> To deep dive into the removal, refer to this [guide](https://www.jenkins.io/blog/2023/06/20/remove-outdated-plugins-while-using-docker/).

Reviewer 2: This section is confusing as it mixes different cases without explaining why and what.

Again, this should not be a step by step guide: if you feel like writing a step by step guide, than jenkins.io's documentation would be a better location to write so

WDYT about:

- Starting by explaining that the already installed plugins are living in a Docker volume which has a different lifecycle than the container image and the container image's entrypoint does **NOT** clean up automatically plugins
- Then describe a command line based scenario to remove plugins manually in the volume (which is roughly the step 14. in your writing here) so it would apply to everyone
  - Mentions (using an [adminition](https://github.com/orgs/community/discussions/16925)) that the UI could be used but it does not properly removes every files associated to `<plugin name>` installation
  -Updates the instructions as you want user to remove `/var/jenkins_home/plugins/<plugin name>*` so that the plugin folder AND all the `*.jpi`, `*.hpi` and any lock files are deleted for the given plugin `<plugin name>`

**To remove a plugin from Jenkins using a custom Dockerfile, you can follow these steps:**

- Create a file called `plugins.txt` in the root directory of your Jenkins project. This file should contain the short names of the plugins you want to remove, one plugin per line.
- In your Dockerfile, add the following command to copy the plugins.txt file into the Jenkins container:

`COPY plugins.txt /usr/share/jenkins/ref/`

- Add the following command to your Dockerfile to remove the plugins specified in the `plugins.txt` file:

`RUN /usr/local/bin/install-plugins.sh < /usr/share/jenkins/ref/plugins.txt`

- Build your Docker image and run the container. The plugins specified in the plugins.txt file will be removed from Jenkins.

Here is an example of a Dockerfile that can be used to remove the "WMI" windows-slave plugin from Jenkins:

Create the custom Dockerfile:

```DockerFile
FROM jenkins/jenkins:latest

# Copy the plugins.txt file into the Jenkins container
COPY plugins.txt /usr/share/jenkins/ref/

# Remove the plugins specified in the plugins.txt file

> RUN /usr/local/bin/install-plugins.sh < /usr/share/jenkins/ref/plugins.txt

Reviewer 3: As far as I know, this does not work with the current Jenkins container images.  The current container images have `install-plugins.sh` as a small script that warns the script has been removed.  The container images include a replacement script `/bin/jenkins-plugin-cli`.  However, it takes a `plugins.txt` file that lists the plugins to be installed, not the plugins to be removed.

I don't think that the `plugins.txt` should be used to define the list of plugins to remove, since it is used in every other location in the container images as the list of plugins to retain.
```suggestion
RUN jenkins-plugin-cli --clean-download-directory
```

Author: Addressed in https://github.com/jenkinsci/docker/pull/1538/commits/b0b6bcf1363f984983d02f4208cb8ffe3e6a9183

To completely remove the plugin from Jenkins, you will need to delete the plugin files from the `$JENKINS_HOME/plugins` directory and restart the Jenkins container. You can do this by manually deleting the files and restarting the container using the command line.

**To remove the plugin files using the command line, follow these steps:**

- Connect to the Jenkins server using the ssh command.
- Navigate to the `$JENKINS_HOME/plugins/` directory using the `cd $JENKINS_HOME/plugins/` command.
- Find the plugin files you want to delete. The plugin files are typically named after the short name of the plugin, with a `.hpi` or `.jpi` file extension.
- Use the `rm` command to delete the plugin files: `rm PLUGIN_SHORT_NAME.hpi`
- Restart the Jenkins container to apply the changes: `docker restart jenkins`

Once the plugin files have been deleted and the Jenkins container has been restarted, the plugin will be completely removed from Jenkins.

> ## Self-Diagonosis

Reviewer: ```suggestion
## Self-Diagnosis
```

Author: Addressed in https://github.com/jenkinsci/docker/pull/1538/commits/033ac90efac6e8fda1f3e388f3729943ee06b4c6

Otherwise a banner will appear prompting the user to install additional plugins,
which may be inappropriate.

## Uninstalling Plugins

**To uninstall a plugin in Jenkins using the web UI, follow these steps:**

- Open Jenkins in your web browser and log in as an administrator.
- Click on the "Manage Jenkins" link in the left menu.
- Click on the "Manage Plugins" link.

> - In the "Installed" tab, find the plugin you want to uninstall and click on the checkbox next to it.

Reviewer: ```suggestion
- In the "Installed Plugins" tab, find the plugin you want to uninstall and click on the checkbox next to it.
```

> - Click on the "Uninstall" button at the bottom of the page.

Reviewer: ```suggestion
```
I haven't seen any "Uninstall" button.

- A message will appear asking you to confirm the uninstallation. Click on the "Yes" button to confirm.

> Jenkins will uninstall the plugin and restart the server to apply the changes. Once the server has restarted, the plugin will be uninstalled and will no longer be available in Jenkins.

Reviewer 2: This assertion is wrong in the context of a Jenkins container. Please try it yourself to see the result.

**To uninstall a plugin in Jenkins using the CLI, follow these steps:**

- Open a terminal window and connect to your Jenkins server using the ssh command. For example:

> `ssh jenkins@your-server-name`

Reviewer 2: The section "uninstall with the CLI" should not be described here. As much as it would make sense in the general jenkins.io documentation, it does not make sense at all in the context of the Docker image:

- There is no SSH at all in Docker image for the controller and it should not even be mentionned here.
- The authentication with a token is required as explained by @gounthar 
- We target configuration as code and immutability: it's not what this chapter shows.

- Once you are logged in to the Jenkins server, enter the following command to list all the installed plugins:

`java -jar jenkins-cli.jar -s http://localhost:8080/ list-plugins`

> - Find the short name of the plugin you want to uninstall from the list of installed plugins. The short name is the name of the plugin as it appears in the `.hpi` or `.jpi` file, without the file extension.

Reviewer: ```suggestion
- Find the short name of the plugin you want to uninstall from the list of installed plugins. It is the one appearing in the first column. For example, for the Jakarta Mail API plugin, you would find `jakarta-mail-api`. It is the root of the plugin file name as it appears in the `.hpi` or `.jpi` binary file (the compiled plugin).
```

- To uninstall the plugin, enter the following command, replacing `PLUGIN_SHORT_NAME` with the short name of the plugin you want to uninstall:

> `java -jar jenkins-cli.jar -s http://localhost:8080/ uninstall-plugin PLUGIN_SHORT_NAME`

Reviewer: Unfortunately, the `uninstall` command does not exist, or does not exist anymore.
The only thing that vaguely looks like it is `disable-plugin` which does not uninstall as far as I know.

Author: Could you share me which command is available then?

Reviewer: A command to uninstall a plugin?
I haven't found any yet. The one which looks the closest is `disable-plugin`, but I'm not sure it will get rid of the plugin on top of disabling it.
Here is the list I get with the latest weekly:

```bash
 add-job-to-view
    Adds jobs to view.
  build
    Builds a job, and optionally waits until its completion.
  cancel-quiet-down
    Cancel the effect of the "quiet-down" command.
  clear-queue
    Clears the build queue.
  connect-node
    Reconnect to a node(s)
  console
    Retrieves console output of a build.
  copy-job
    Copies a job.
  create-credentials-by-xml
    Create Credential by XML
  create-credentials-domain-by-xml
    Create Credentials Domain by XML
  create-job
    Creates a new job by reading stdin as a configuration XML file.
  create-node
    Creates a new node by reading stdin as a XML configuration.
  create-view
    Creates a new view by reading stdin as a XML configuration.
  declarative-linter
    Validate a Jenkinsfile containing a Declarative Pipeline
  delete-builds
    Deletes build record(s).
  delete-credentials
    Delete a Credential
  delete-credentials-domain
    Delete a Credentials Domain
  delete-job
    Deletes job(s).
  delete-node
    Deletes node(s)
  delete-view
    Deletes view(s).
  disable-job
    Disables a job.
  disable-plugin
    Disable one or more installed plugins.
  disconnect-node
    Disconnects from a node.
  enable-job
    Enables a job.
  enable-plugin
    Enables one or more installed plugins transitively.
  get-credentials-as-xml
    Get a Credentials as XML (secrets redacted)
  get-credentials-domain-as-xml
    Get a Credentials Domain as XML
  get-gradle
    List available gradle installations
  get-job
    Dumps the job definition XML to stdout.
  get-node
    Dumps the node definition XML to stdout.
  get-view
    Dumps the view definition XML to stdout.
  groovy
    Executes the specified Groovy script.
  groovysh
    Runs an interactive groovy shell.
  help
    Lists all the available commands or a detailed description of single command.
  import-credentials-as-xml
    Import credentials as XML. The output of "list-credentials-as-xml" can be used as input here as is, the only needed change is to set the actual Secrets which are redacted in the output.
  install-plugin
    Installs a plugin either from a file, an URL, or from update center.
  keep-build
    Mark the build to keep the build forever.
  list-changes
    Dumps the changelog for the specified build(s).
  list-credentials
    Lists the Credentials in a specific Store
  list-credentials-as-xml
    Export credentials as XML. The output of this command can be used as input for "import-credentials-as-xml" as is, the only needed change is to set the actual Secrets which are redacted in the output.
  list-credentials-context-resolvers
    List Credentials Context Resolvers
  list-credentials-providers
    List Credentials Providers
  list-jobs
    Lists all jobs in a specific view or item group.
  list-plugins
    Outputs a list of installed plugins.
  mail
    Reads stdin and sends that out as an e-mail.
  offline-node
    Stop using a node for performing builds temporarily, until the next "online-node" command.
  online-node
    Resume using a node for performing builds, to cancel out the earlier "offline-node" command.
  quiet-down
    Quiet down Jenkins, in preparation for a restart. Don’t start any builds.
  reload-configuration
    Discard all the loaded data in memory and reload everything from file system. Useful when you modified config files directly on disk.
  reload-job
    Reload job(s)
  remove-job-from-view
    Removes jobs from view.
  replay-pipeline
    Replay a Pipeline build with edited script taken from standard input
  restart
    Restart Jenkins.
  restart-from-stage
    Restart a completed Declarative Pipeline build from a given stage.
  safe-restart
    Safely restart Jenkins.
  safe-shutdown
    Puts Jenkins into the quiet mode, wait for existing builds to be completed, and then shut down Jenkins.
  session-id
    Outputs the session ID, which changes every time Jenkins restarts.
  set-build-description
    Sets the description of a build.
  set-build-display-name
    Sets the displayName of a build.
  shutdown
    Immediately shuts down Jenkins server.
  stop-builds
    Stop all running builds for job(s)
  update-credentials-by-xml
    Update Credentials by XML
  update-credentials-domain-by-xml
    Update Credentials Domain by XML
  update-job
    Updates the job definition XML from stdin. The opposite of the get-job command.
  update-node
    Updates the node definition XML from stdin. The opposite of the get-node command.
  update-view
    Updates the view definition XML from stdin. The opposite of the get-view command.
  version
    Outputs the current version.
  wait-node-offline
    Wait for a node to become offline.
  wait-node-online
    Wait for a node to become online.
  who-am-i
    Reports your credential and permissions.
```

- Jenkins will uninstall the plugin and restart the server to apply the changes. Once the server has restarted, the plugin will be uninstalled and will no longer be available in Jenkins.

**Note**: Some plugins may have dependencies on other plugins, which means that uninstalling one plugin may also uninstall other plugins that depend on it. Make sure to carefully review the list of plugins that will be uninstalled before confirming the uninstallation.

To check the dependencies, use the following command:

> `java -jar jenkins-cli.jar -s http://localhost:8080/ list-dependencies PLUGIN_SHORT_NAME`

Reviewer: Same for this one, it does not seem to exist anymore.

Author: Any way to list dependencies currently?

Reviewer: I think it's gone. 😢

Replace `PLUGIN_SHORT_NAME` with the short name of the plugin you want to check the dependencies for. The command will output a list of the plugins that the selected plugin is dependent on.

**To remove a plugin from Jenkins using a custom Dockerfile, you can follow these steps:**

- While installing the plugins, you will need to create a file called `plugins.txt` in the root directory of your Jenkins project. Remove the plugin name from this file which you wish to uninstall.
- In your Dockerfile, add the following command to copy the plugins.txt file into the Jenkins container:

> 

Reviewer: I'm afraid this won't be enough.
I've been there and done that, and it was not sufficient.
Let's say it's ok for the time being, and I will make another PR later on when I'll have a clear idea on how to remove a plugin when using Docker.

Author: I have followed through [these steps ](https://github.com/jenkinsci/docker/pull/1538#issuecomment-1366839208)

Also, could you share me some video with uninstallation process, i think i am missing something here, a screen recorded video can be helpful.

Author: @gounthar

Reviewer: I'll do my best to record that for you asap.
Thanks a lot for your contribution.

Author: @gounthar any update

Create the custom Dockerfile:

```DockerFile
FROM jenkins/jenkins:latest

# Copy the plugins.txt file into the Jenkins container
COPY --chown=jenkins:jenkins plugins.txt $REF/plugins.txt

# Remove the plugins not exist in plugins.txt file
RUN jenkins-plugin-cli --plugin-file $REF/plugins.txt
```

In this example, update the `plugins.txt` file by removing the following line:

`wmi-windows-slaves`

When you build the Docker image and run the container, the "WMI" windows-slave plugin will be removed from Jenkins.

Note that when you uninstall a plugin from Jenkins, the plugin files are not deleted from the `$JENKINS_HOME/plugins/` directory by default. This is because the `$JENKINS_HOME` directory is typically used as a data volume or bind mounted volume, and deleting the plugin files could potentially cause data loss or other issues.

To completely remove the plugin from Jenkins, you will need to delete the pluginspecified in the files from the `$JENKINS_HOME/plugins` directory and restart the Jenkins container. You can do this by manually deleting the files and restarting the container using the command line.

> **To remove the plugin files using the command line, follow these steps:**
> 
> - Connect to the Jenkins server using the ssh command.
> - Navigate to the `$JENKINS_HOME/plugins/` directory using the `cd $JENKINS_HOME/plugins/` command.
> - Find the plugin files you want to delete. The plugin files are typically named after the short name of the plugin, with a `.hpi` or `.jpi` file extension.
> - Use the `rm` command to delete the plugin files: `rm PLUGIN_SHORT_NAME.hpi`
> - Restart the Jenkins container to apply the changes: `docker restart jenkins`
> 
> Once the plugin files have been deleted and the Jenkins container has been restarted, the plugin will be completely removed from Jenkins.

Reviewer 2: ```suggestion
```

This whole sections duplicates what was written previsously

**To remove a plugin from a Jenkins instance that is using a data volume to preserve the configuration, follow these steps:**

- Remove the plugin from the `plugins.txt` file.
- Rebuild the Docker image using the updated `plugins.txt` file.
- Stop the existing Jenkins container.
- Remove the plugin files from the `$JENKINS_HOME/plugins/` directory in the data volume.
- You can do this by attaching the data volume to another container and deleting the plugin files directly, or by using a tool such as `rsync` to delete the files from the host machine.
- Start the Jenkins container using the updated Docker image.

This process will ensure that the Jenkins instance is using the updated set of plugins specified in the `plugins.txt` file.

## Self-Diagnosis

To self-diagnose the plugin removal step in your Jenkins setup, you can follow these steps:

> - Connect to the Jenkins server using the command line interface (CLI). This can be done using the ssh command, for example:
> 
> `ssh jenkins@your-server-name`
> 
> - Once you are logged in to the Jenkins server, you can use the following command to list all the installed plugins:
> 
> `java -jar jenkins-cli.jar -s http://localhost:8080/ list-plugins`

Reviewer 2: ```suggestion
```

Does not apply to a Docker container
